massa-versioning-worker/src/versioning.rs

```diff
@@ -0,0 +1,1074 @@
+use std::cmp::Ordering;
+use std::collections::{BTreeMap, BTreeSet, HashMap};
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+use std::sync::Arc;
+
+use machine::{machine, transitions};
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use parking_lot::RwLock;
+use thiserror::Error;
+
+use massa_models::{amount::Amount, config::VERSIONING_THRESHOLD_TRANSITION_ACCEPTED};
+use massa_time::MassaTime;
+
+// TODO: add more items here
+/// Versioning component enum
+#[allow(missing_docs)]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum MipComponent {
+    Address,
+    Block,
+    VM,
+}
+
+/// MIP info (name,  version,
+#[derive(Clone, Debug)]
+pub struct MipInfo {
+    /// MIP name or descriptive name
+    pub name: String,
+    /// Network (or global) version (to be included in block header)
+    pub version: u32,
+    /// Component concerned by this versioning (e.g. a new Block version)
+    pub component: MipComponent,
+    /// Component version
+    pub component_version: u32,
+    /// a timestamp at which the version gains its meaning (e.g. accepted in block header)
+    pub start: MassaTime,
+    /// a timestamp at which the deployment is considered active or failed (timeout > start)
+    pub timeout: MassaTime,
+}
+
+// Need Ord / PartialOrd so it is properly sorted in BTreeMap
+
+impl Ord for MipInfo {
+    fn cmp(&self, other: &Self) -> Ordering {
+        (self.start, &self.timeout).cmp(&(other.start, &other.timeout))
+    }
+}
+
+impl PartialOrd for MipInfo {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for MipInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+            && self.version == other.version
+            && self.component == other.component
+            && self.start == other.start
+            && self.timeout == other.timeout
+    }
+}
+
+impl Eq for MipInfo {}
+
+// Need to impl this manually otherwise clippy is angry :-P
+impl Hash for MipInfo {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.name.hash(state);
+        self.version.hash(state);
+        self.component.hash(state);
+        self.start.hash(state);
+        self.timeout.hash(state);
+    }
+}
+
+machine!(
+    /// State machine for a Versioning component that tracks the deployment state
+    #[allow(missing_docs)]
+    #[derive(Clone, Copy, Debug, PartialEq)]
+    enum MipState {
```

Comment:
can you explain the meaning of "MIP" in comment :innocent:

Reply:
@modship can you check the doc in lib.rs and report? :)